src/rpc/mining.cpp

```diff
@@ -390,7 +390,7 @@ static RPCHelpMan generateblock()
         LOCK(cs_main);
```

Comment:
It is also acquired within `miner.testBlockValidity` (and in `miner.createNewBlock`) so maybe is now unnecessary outside the Miner interface?

https://github.com/bitcoin/bitcoin/blob/a9716c53f05082d6d89ebea51a46d4404efb12d7/src/node/interfaces.cpp#L873-L877

Reply:
I pushed 1f0559e9954253c23af7ee9b49e57eb3d7a45bea to drop it from `createNewBlock`. Will look at the other occurances too.

Reply:
The lock at line 390 has been there from the introduction of the `generateblock` RPC call in #17693. I don't see any discussion of it in the reviews.

I think it's because `assert(pindexPrev && pindexPrev == chainstate.m_chain.Tip());` inside of `TestBlockValidity`. Added d23968e0fb6e1d7c5d19b07722796181d6e04f5b to explain that, and also make `testBlockValidity` require the lock.

Reply:
Actually that assert isn't the problem, it merely checks that the `chainman().ActiveChainstate()` and `chainman().ActiveChain().Tip()` that we passed in match.

The problem is deeper down the call stack, where `TestBlockValidity` calls `ConnectBlock`. That takes a `viewNew` argument which is `chainstate.CoinsTip()`, and our proposed `block`, and then it does `assert(hashPrevBlock == view.GetBestBlock());`.

Reply:
Pushed c8343e29a66082b37c37ee0ff8f7433b97eea9b1 to deal with this (replaces https://github.com/bitcoin/bitcoin/commit/18d6b0f86769f89738a3b08aa8152a8479cef5d4). This drops both locks from the RPC code.

An alternative that would actually prevent the tip from updating, rather than fail the RPC call, is to hold the lock even longer. But callers need to handle failure anyway since `testBlockValidity` can fail for other reasons.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30335#discussion_r1654338332

> What is this lock used for?

Marking this thread resolved as the lock is now removed in later commit c8343e29a66082b37c37ee0ff8f7433b97eea9b1